### PR TITLE
Feature/develop2 graph complete

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Feature: Implement the new graph model for 2.0 proposals https://github.com/conan-io/tribe/pull/27, https://github.com/conan-io/tribe/pull/26, https://github.com/conan-io/tribe/pull/20
Docs: Omit

### CODE REVIEW

File: conans/client/graph/graph_manager.py [hunk 1]
```diff
@@ -81,11 +50,22 @@ def load_graph(self, reference, create_reference, profile_host, profile_build, g
         root_node = self._load_root_node(reference, create_reference, profile_host, graph_lock,
                                          root_ref, lockfile_node_id, is_build_require,
                                          require_overrides)
-        deps_graph = self._resolve_graph(root_node, profile_host, profile_build, graph_lock,
-                                         build_mode, check_updates, update, remotes, recorder,
-                                         apply_build_requires=apply_build_requires)
-        # Run some validations once the graph is built
-        self._validate_graph_provides(deps_graph)
+        profile_host_build_requires = profile_host.build_requires
```
Comment (czoido) [inline conans/client/graph/graph_manager.py:53]:
```suggestion
```

File: conans/client/graph/graph.py [hunk 2]
```diff
@@ -66,39 +44,108 @@ def __init__(self, ref, conanfile, context, recipe=None, path=None):
...
+
+    def propagate_closing_loop(self, require, prev_node):
+        self.propagate_downstream(require, prev_node)
+        # List to avoid mutating the dict
+        for transitive in list(prev_node.transitive_deps.values()):
+            # TODO: possibly optimize in a bulk propagate
+            prev_node.propagate_downstream(transitive.require, transitive.node, self)
+
+    def propagate_downstream(self, require, node, src_node=None):
+        # print("  Propagating downstream ", self, "<-", require)
```
Comment (czoido) [inline conans/client/graph/graph.py:79]:
```suggestion
```

File: conans/client/graph/graph.py [hunk 3]
```diff
@@ -66,39 +44,108 @@ def __init__(self, ref, conanfile, context, recipe=None, path=None):
...
+        assert node is not None
+        # This sets the transitive_deps node if it was None (overrides)
+        # Take into account that while propagating we can find RUNTIME shared conflicts we
+        # didn't find at check_downstream_exist, because we didn't know the shared/static
+        existing = self.transitive_deps.get(require)
+        if existing is not None and existing.require is not require:
+            if existing.node is not None and existing.node.ref != node.ref:
+                # print("  +++++Runtime conflict!", require, "with", node.ref)
```
Comment (czoido) [inline conans/client/graph/graph.py:87]:
```suggestion
```

File: conans/client/graph/graph.py [hunk 4]
```diff
@@ -66,39 +44,108 @@ def __init__(self, ref, conanfile, context, recipe=None, path=None):
...
+        # TODO: Make self.ref always exist, but with name=None if name not defined
+        if self.ref is not None and require.ref.name == self.ref.name:
+            if require.build and require.ref.version != self.ref.version:
+                pass  # Allow bootstrapping
+            else:
+                return None, self, self  # First is the require, as it is a loop => None
+
+        # First do a check against the current node dependencies
+        prev = self.transitive_deps.get(require)
+        # print("    Transitive deps", self.transitive_deps)
```
Comment (czoido) [inline conans/client/graph/graph.py:124]:
```suggestion
```

File: conans/client/graph/graph.py [hunk 5]
```diff
@@ -66,39 +44,108 @@ def __init__(self, ref, conanfile, context, recipe=None, path=None):
...
+
+        # Seems the algrithm depth-first, would only have 1 dependant at most to propagate down
+        # at any given time
+        if not self.dependants:
+            return
+        assert len(self.dependants) == 1
+        d = self.dependants[0]
+
+        # TODO: Implement an optimization where the requires is checked against a graph global
+        # print("    Lets check_downstream one more")
```
Comment (czoido) [inline conans/client/graph/graph.py:141]:
```suggestion
```

File: conans/client/graph/graph.py [hunk 6]
```diff
@@ -66,39 +44,108 @@ def __init__(self, ref, conanfile, context, recipe=None, path=None):
...
+        down_require = d.require.transform_downstream(self.conanfile.package_type, require, None)
+
+        if down_require is None:
+            # print("    No need to check dowstream more")
```
Comment (czoido) [inline conans/client/graph/graph.py:145]:
```suggestion
```

File: conans/client/graph/graph_binaries.py [hunk 7]
```diff
@@ -257,128 +257,17 @@ def _process_node(self, node, pref, build_mode, update, remotes):
...
-            conanfile.options.validate()
-        except ConanException:
-            conanfile.info.invalid = BINARY_INVALID, str(e)
-
-        info = conanfile.info
-        node.package_id = info.package_id()
+    def _evaluate_package_id(self, node):
+        compute_package_id(node, self._cache.new_config)  # TODO: revise compute_package_id()
 
     def evaluate_graph(self, deps_graph, build_mode, update, remotes, nodes_subset=None, root=None):
```
Comment (czoido) [inline conans/client/graph/graph_binaries.py:263]:
```suggestion
    def evaluate_graph(self, deps_graph, build_mode, update, remotes, nodes_subset=None):
```

File: conans/client/graph/graph_binaries.py [hunk 8]
```diff
@@ -257,128 +257,17 @@ def _process_node(self, node, pref, build_mode, update, remotes):
...
+        build_mode = BuildMode(build_mode, self._output)
+        assert isinstance(build_mode, BuildMode)
+
         default_package_id_mode = self._cache.config.default_package_id_mode
         default_python_requires_id_mode = self._cache.config.default_python_requires_id_mode
```
Comment (czoido) [inline conans/client/graph/graph_binaries.py:268]:
```suggestion
```

File: conans/client/graph/graph_manager.py [hunk 9]
```diff
@@ -1,49 +1,18 @@
...
-        context = CONTEXT_HOST if force_host_context else self._default_context
-        self.add(build_require, context, force_host_context)
-
-    def __str__(self):
-        items = ["{} ({})".format(br, ctxt) for (_, ctxt), br in self.items()]
-        return ", ".join(items)
-
-
 class GraphManager(object):
     def __init__(self, output, cache, remote_manager, loader, proxy, resolver, binary_analyzer):
```
Comment (czoido) [inline conans/client/graph/graph_manager.py:11]:
```suggestion
    def __init__(self, output, cache, loader, proxy, resolver, binary_analyzer):
```

File: conans/client/graph/graph.py (hunk 2 above)
Comment (memsharded) [reply conans/client/graph/graph.py:79]: These have been extraordinary useful for debugging things, I will surely remove them, but please let them stay a bit longer, at least just before merging this.

File: conans/client/graph/graph_binaries.py (hunk 8 above)
Comment (memsharded) [reply conans/client/graph/graph_binaries.py:268]: Better add FIXMES, at least python_requires_id_mode will partially remain, and we need to talk what to do with the package_id.

**Review by czoido — APPROVED** [review]

**Pull request merged**